mirrord/intproxy/src/proxies/incoming.rs

```diff
@@ -580,3 +568,87 @@ impl BackgroundTask for IncomingProxy {
         }
     }
 }
+
+impl IncomingProxy {
+    /// Sends back the streamed http response to the agent.
+    ///
+    /// If we cannot get the next frame of the streamed body, then we retry the whole
+    /// process, by sending the original `request` again through the http `interceptor` to
+    /// our hyper handler.
+    #[allow(clippy::type_complexity)]
+    #[tracing::instrument(level = Level::TRACE, skip(self), ret)]
+    async fn streamed_http_response(
+        &mut self,
+        mut response: HttpResponse<StreamBody<ReceiverStream<Result<Frame<Bytes>, hyper::Error>>>>,
+        request: Option<HttpRequestFallback>,
+    ) -> Option<ClientMessage> {
+        let mut body = vec![];
+        let key = (response.connection_id, response.request_id);
+
+        match response
+            .internal_response
+            .body
+            .next_frames(false)
+            .await
+            .map_err(InterceptorError::from)
+        {
+            Ok(frames) => {
+                frames
+                    .frames
+                    .into_iter()
+                    .map(From::from)
+                    .for_each(|frame| body.push(frame));
+
+                self.response_body_rxs
+                    .insert(key, StreamNotifyClose::new(response.internal_response.body));
+
+                let internal_response = InternalHttpResponse {
+                    status: response.internal_response.status,
+                    version: response.internal_response.version,
+                    headers: response.internal_response.headers,
+                    body,
+                };
+                let response = ChunkedResponse::Start(HttpResponse {
+                    port: response.port,
+                    connection_id: response.connection_id,
+                    request_id: response.request_id,
+                    internal_response,
+                });
+                Some(ClientMessage::TcpSteal(LayerTcpSteal::HttpResponseChunked(
+                    response,
+                )))
+            }
+            // Retry on `RST_STREAM` error.
+            Err(InterceptorError::Reset) => {
+                tracing::warn!("`RST_STREAM` received in the response, retrying!");
```

Comment:
```suggestion
            // Retry on known errors.
            Err(error @ InterceptorError::Reset) | Err(error @ InterceptorError::ConnectionClosedTooSoon(..)) => {
                tracing::warn!(%error, ?request, "Failed to read first frames of streaming HTTP response");
```
For consistency with interceptor check